keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md

```diff
@@ -0,0 +1,398 @@
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Provide an interface so that users can implement a custom queue of scheduler.
+With such, users can build their custom queue implementation at will, and plumb into
+the scheduler seamlessly. This can satisfy the business needs such as sophisticated
+pods sorting and internal queuing mechanics.
+
+## Motivation
+
+The current internal queue implementation ([scheduling_queue.go#PriorityQueue](https://github.com/kubernetes/kubernetes/blob/ee459b8969ed2abfed79a07d4ac9d41f13f18ce6/pkg/scheduler/internal/queue/scheduling_queue.go#L126))
+works for most cases, but the limited interfaces exposure makes it hard to
+implement a scheduler plugin for sophisticated requirements.
+
+One requirement in terms of multi-tenancy support:  
+There are many pods from 2 users (userA and userB), and the target is to ensure resource
+usage (E.g. userA gets x CPUs) ratio between userA and userB is 1:1. Sorting the pods by
+the resource usage in `Less` function doesn't work properly, because the resource usage for
+userA/userB will be updated dynamically once a pod is selected and bound, while the
+algorithm (heapsort) that used by the current queue to get the next pods depends on static
+data.
+
+Other requirements come from current semi-exposed functions/parameters:
+For example, the interval of function flushUnschedulableQLeftover, although configurable,
+is immutable after it gets initialized. A requirement is to adjust the interval dynamically
+based on the pressure of pending workloads (e.g., length of queue or other metrics).
+
+Given that the business needs may vary greatly, it'd be desirable to provide a replaceable
+`SchedulerQueue` interface and plumbing mechanics. So that the developers can
+implement their custom queue implementation, while the upstream maintainers focus
+on keeping the core small and extensible, and thus only maintain the internal queue piece.
+
+### Goals
+
+1. Support custom queue implementation of the scheduler.
+2. Roll out the design gradually. In phase 1, keep current internal queue implementation
+intact.
+
+### Non-Goals
+
+Support custom rules for pod selection in the current internal queue of scheduler.
+
+## Proposal
+
+Provide an interface like current scheduler plugins, users can provide a
+custom queue of scheduler and register it, then the kubernetes
```

Comment:
at which time? build time, configuration time, runtime?

Reply:
At build time, the custom queue will be part of the scheduler binary after building.  Readme updated.